etc/vcl_snippets/deliver.vcl

```diff
@@ -8,10 +8,20 @@
         set resp.http.Cache-Control = "no-store, no-cache, must-revalidate, max-age=0";
     }
 
+    # GraphQl queries should vary on X-Magento-Cache-Id if it was supplied
+    if ( req.http.graphql && resp.http.X-Magento-Cache-Id ) {
+        set resp.http.Vary = regsub(resp.http.Vary, "(?i)X-Magento-Vary,Https(,X-Magento-Cache-Id)?", "X-Magento-Cache-Id");
+    }
+
     # Execute only on the edge nodes
     if ( fastly.ff.visits_this_service == 0 ) {
-        # Remove X-Magento-Vary and HTTPs Vary served to the user
-        set resp.http.Vary = regsub(resp.http.Vary, "(?i)X-Magento-Vary,Https", "Cookie");
+        if ( req.http.graphql ) {
+            # GraphQl queries without X-Magento-Cache-Id should be cached locally under store and currency
+            set resp.http.Vary = regsub(resp.http.Vary, "(?i)X-Magento-Vary,Https(,X-Magento-Cache-Id)?", "Authentication,Store,Content-Currency");
```

Comment:
@cpartica Question: Does the Authentication header need to be considered in local cache as well?

Reply:
It's not really possible to change cache key in response without going through a varnish restart. Unfortunately that would result in throwing away the response.

Couple concerns about this. Varnish only supports up to 200 different variations per cache key so if you will be having a lot more then that you should mutate the cache key. 

I am still unclear how the X-Magento-Cache-Id header is supposed to work.

Reply:
hi @vvuksan . It will be better to connect on Magento's slack (magentocommeng.slack.com), we left you some messages there.
Having said that, the X-Magento-Cache-Id will work similar to how Luma works with X-Magento-Vary cookie on dealing with cache variations.
Indeed we intentionally skipped caching content when sending Authentication header, but that's about to change.
We have some diagrams that we can show you to better understand how X-Magento-Cache-Id works, though it's exactly like X-Magento-Vary cookie from Luma, only throgh a header.
We think that if luma is supported through that algorithm, this will  be too.